### PR TITLE
block pool simulator

### PR DESCRIPTION
**Pull request by arnetheduck**
like state_sim, but more

### CODE REVIEW

File: research/simutils.nim [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import
```
Comment (tersec) [inline research/simutils.nim:1]: This refactoring should be helpful as well in allowing other sort of parameterized simulations and clarifying the flow of the main simulation modules by extracting the utility functions.

File: research/simutils.nim [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  stats, os, strformat,
+  ../tests/[testblockutil],
+  ../beacon_chain/[extras, ssz],
+  ../beacon_chain/spec/[beaconstate, datatypes, digest, helpers]
+
+template withTimer*(stats: var RunningStat, body: untyped) =
```
Comment (tersec) [inline research/simutils.nim:7]: `withTimer` and `withTimerRet` have popped up in `nim-beacon-chain` in various guises in at least 2 or 3 places, near-identically. I started using `monotimes` at some point, but I'm not so sure I like its interface -- it's got less than ideal units with which to work, even if conceptually it's the right thing. So maybe a TODO would be to merge these as much as feasible -- only so many `let foo = time(); doThing(); show(time() - foo)` one needs, but very much shouldn't block merging.

File: tests/testblockutil.nim [hunk 3]
```diff
@@ -25,7 +25,7 @@ func makeFakeHash(i: int): Eth2Digest =
   static: doAssert sizeof(bytes) <= sizeof(result.data)
   copyMem(addr result.data[0], addr bytes[0], sizeof(bytes))
 
-func hackPrivKey(v: Validator): ValidatorPrivKey =
+func hackPrivKey*(v: Validator): ValidatorPrivKey =
```
Comment (tersec) [inline tests/testblockutil.nim:28]: Not obvious to me at a glance why the other changes would have forced the scope of this to become global?

Comment (arnetheduck) [reply tests/testblockutil.nim:28]: https://github.com/status-im/nim-beacon-chain/pull/956/files#diff-c66bb19c940b38a856d465a5e4a3d8b9R81

File: research/simutils.nim (hunk 2 above)
Comment (arnetheduck) [inline research/simutils.nim:7]:
```suggestion
template withTimer*(stats: var RunningStat, body: untyped) =
  # TODO unify timing somehow
```

**Pull request merged**